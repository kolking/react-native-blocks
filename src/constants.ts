export const Colors = {
  white: '#FFFFFF',
  black: '#030303',
  navyBlue: '#3E88C5',
};
