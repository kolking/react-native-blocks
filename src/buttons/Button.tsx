import React from 'react';
import {
  Image,
  ImageStyle,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  ImageSourcePropType,
  View,
  ViewStyle,
} from 'react-native';

import { Colors } from '../constants';

export interface Props extends TouchableOpacityProps {
  testID?: string;
  visible?: boolean;
  children?: React.ReactNode;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  iconSource?: ImageSourcePropType;
  iconStyle?: StyleProp<ImageStyle>;
}

const Button: React.FC<Props> = ({
  visible,
  children,
  style,
  textStyle,
  iconSource,
  iconStyle,
  ...props
}) => {
  return !visible ? null : (
    <TouchableOpacity testID="button" activeOpacity={0.5} {...props}>
      <View style={[styles.button, style, props.disabled && styles.disabled]}>
        {children && (
          <Text style={[styles.buttonText, textStyle]} numberOfLines={1}>
            {children}
          </Text>
        )}
        {iconSource && <Image style={[styles.buttonIcon, iconStyle]} source={iconSource} />}
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    height: 50,
    borderRadius: 5,
    alignItems: 'center',
    flexDirection: 'row',
    paddingHorizontal: 15,
    justifyContent: 'center',
    backgroundColor: Colors.navyBlue,
  },
  buttonText: {
    height: 16,
    fontSize: 16,
    lineHeight: 16,
    color: Colors.white,
    textTransform: 'uppercase',
  },
  buttonIcon: {
    width: 20,
    height: 20,
    flexShrink: 0,
    tintColor: Colors.white,
  },
  disabled: {
    opacity: 0.5,
  },
});

Button.defaultProps = {
  visible: true,
  disabled: false,
};

export default Button;
