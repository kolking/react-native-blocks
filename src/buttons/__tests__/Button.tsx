import React from 'react';
import { Text } from 'react-native';

// Note: test renderer must be required after react-native.
import TestRenderer from 'react-test-renderer';

import { Colors } from '../../constants';

/**
 * Under test
 */
import Button, { Props } from '../Button';

jest.mock('react-native/Libraries/Components/Touchable/TouchableOpacity', () => 'TouchableOpacity');

const createElement = (props: Props) => (
  <Button {...props}>
    <Text>TEXT</Text>
  </Button>
);

const createRenderer = (props: Props) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  expect(createRenderer({})).toMatchSnapshot();
});

it('should render as disabled', () => {
  expect(createRenderer({ disabled: true })).toMatchSnapshot();
});

it('should not render', () => {
  expect(createRenderer({ visible: false })).toMatchSnapshot();
});

it('should render with custom styles', () => {
  expect(
    createRenderer({
      style: { backgroundColor: Colors.black },
      textStyle: { color: Colors.white },
    }),
  ).toMatchSnapshot();
});

it('should render with icon', () => {
  expect(
    createRenderer({ iconSource: { uri: 'image.png' }, iconStyle: { tintColor: Colors.white } }),
  ).toMatchSnapshot();
});

it('should call onPress', () => {
  const onPress = jest.fn();
  const tree = createRenderer({ onPress });

  const button = tree.root.findByProps({ testID: 'button' });

  button.props.onPress();

  expect(onPress).toBeCalled();
});
