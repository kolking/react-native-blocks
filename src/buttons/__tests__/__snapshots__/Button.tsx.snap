// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render 1`] = `null`;

exports[`should render as disabled 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  disabled={true}
  testID="button"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#3E88C5",
          "borderRadius": 5,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 15,
        },
        undefined,
        Object {
          "opacity": 0.5,
        },
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "height": 16,
            "lineHeight": 16,
            "textTransform": "uppercase",
          },
          undefined,
        ]
      }
    >
      <Text>
        TEXT
      </Text>
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`should render normally 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  disabled={false}
  testID="button"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#3E88C5",
          "borderRadius": 5,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 15,
        },
        undefined,
        false,
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "height": 16,
            "lineHeight": 16,
            "textTransform": "uppercase",
          },
          undefined,
        ]
      }
    >
      <Text>
        TEXT
      </Text>
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`should render with custom styles 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  disabled={false}
  testID="button"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#3E88C5",
          "borderRadius": 5,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 15,
        },
        Object {
          "backgroundColor": "#030303",
        },
        false,
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "height": 16,
            "lineHeight": 16,
            "textTransform": "uppercase",
          },
          Object {
            "color": "#FFFFFF",
          },
        ]
      }
    >
      <Text>
        TEXT
      </Text>
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`should render with icon 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  disabled={false}
  testID="button"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#3E88C5",
          "borderRadius": 5,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 15,
        },
        undefined,
        false,
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "height": 16,
            "lineHeight": 16,
            "textTransform": "uppercase",
          },
          undefined,
        ]
      }
    >
      <Text>
        TEXT
      </Text>
    </Text>
    <Image
      source={
        Object {
          "uri": "image.png",
        }
      }
      style={
        Array [
          Object {
            "flexShrink": 0,
            "height": 20,
            "tintColor": "#FFFFFF",
            "width": 20,
          },
          Object {
            "tintColor": "#FFFFFF",
          },
        ]
      }
    />
  </View>
</TouchableOpacity>
`;
